jvm/src/main/kotlin/app/cash/trifle/Certificate.kt

```diff
@@ -28,48 +24,28 @@ data class Certificate internal constructor(
     ).encode()
 
   /**
-   * Verify that the provided certificate matches what we expected.
-   * It matches the CSR that we have and the root cert is what
-   * we expect.
+   * Verify that the provided certificate matches the CSR that we have.
    *
    * @param certificateRequest -  request used to generate this certificate
-   * @param ancestorCertificateChain - list of certificates preceding *this* one.  Namely, the first
-   *   entry should be the certificate corresponding to the issuer of this certificate, and each
-   *   thereafter should be the issuer of the one before it.
-   * @param anchorCertificate - the trust anchor against which we would like to verify the
-   *   ancestorCertificateChain. This may be the terminal (root) certificate of the chain or may be
-   *   an intermediate certificate in the chain which is already trusted.
-   * @param date - The date to use for verification against certificates' validity windows. If null,
-   *   the current time is used.
    *
    * @return - [Result] indicating [Result.isSuccess] or [Result.isFailure]:
    * - success value is expressed as a [Unit] (Nothing)
-   * - failure value is expressed as a [TrifleErrors]
+   * - failure value is expressed as a [TrifleErrors.CSRMismatch] if attributes are mismatched
    */
-  fun verify(
-    certificateRequest: CertificateRequest,
-    ancestorCertificateChain: List<Certificate>,
-    anchorCertificate: Certificate,
-    date: Date? = null
-  ): Result<Unit> {
-    // First check to see if the certificate chain validates
-    val certChainResult = CertChainValidatorFactory.get(anchorCertificate, date)
-      .validate(listOf(this) + ancestorCertificateChain)
-
-    return certChainResult.mapCatching {
-      val x509Certificate = X509CertificateHolder(certificate)
-      when (certificateRequest) {
-        is PKCS10Request -> {
-          // Certificate chain matches, check with certificate request.
-          // TODO(dcashman): Check other attributes as well.
-          if (certificateRequest.pkcs10Req.subject != x509Certificate.subject ||
-            certificateRequest.pkcs10Req.subjectPublicKeyInfo != x509Certificate.subjectPublicKeyInfo
-          ) {
-            throw CSRMismatch
-          }
+  fun verify(certificateRequest: CertificateRequest): Result<Unit> {
```

Comment:
is the `certRequest` required?
it looks like ios doesn't?

```// Validate cert matches the certificate request (so generated key) 
// and the root (so it has been generated by the right CA).
let isValid = certs[0].verify(
            certificateRequest: certRequest,
            intermediateTrifleChain: certs,
            rootTrifleCertificate: root)

// Once it passes validation, certReq is no longer needed and it can be deleted
// Store cert along with the respective keyHandle
        
// To check only for the validity of a stored cert, you can do either of below choices
// Option 1 is a more complete check of the device cert and the full chain
isValid = certs[0].verify(intermediateChain: certs )

// option 2 only checks the validity of the device cert
isValid = certs[0].verify(intermediateChain: [] )
```

Reply:
Might need to re-think about some of the abstractions, but this isn't an API you'd be using here. Rather the surface would be from TrifleApi. This is more of a validity check that is used by the top level API to verify that the attributes from the CertificateRequest matches the newly minted device certificate.

Reply:
@jamesalee213  the only time certRequest is needed is when the app gets the certificate for the first time.
It wants to make sure that the cert it got matches the parameters it had requested.
Once it verifies that, it will throw away the cert request and will no longer have to cross validate the cert with it.
That's why the cert request is optional at the Trifle API level.